### ISSUE TITLE
Include a default prediction

### ISSUE THREAD

**Issue by jcozar87** [scikit-fuzzy/scikit-fuzzy#106]
Now, if an input does not fire any rule, it throws an exception:
`raise ValueError("Crisp output cannot be calculated, likely "...`

Instead, it should return a _default prediction_ (assigned to a ControlSystem through a `set` function), or at least return a `None` value.

What do you think it should be the best option?
In my opinion, return a _default prediction_, but if it has not been set previously return a `None` value.

**Comment by pbreach** [comment]
I haven't gotten to use the package yet for control systems so the weight of my input is limited, but I'm +1 for returning `None` in general, or maybe throwing a custom `ControlSystemError` for explicitly catching the error via try-except instead.

**Comment by JDWarner** [comment]
I'm a little uncomfortable allowing systems to return None, NaN, or similar. It's rather easy to design systems where a single point set of values will break, but the rest of the interval works.

When someone makes a system like this, then tests it, that's easy to miss. Then, when it breaks, it's hard to debug.

I'd be OK with a default value. Ideally this would be optional, and if not provided or set we could make a reasonable guess given the connectivity of the system's terms. Full firing for all connected Terms, take the centroid for example.

**Comment by JDWarner** [comment]
The ultimate solution is likely a check, run when a ControlSystemSimulation is created, which uses the rules' connectivity (and their boolean relationship) to determine if the system is fully connected. Fully connected means that any set of inputs will evaluate to a numerical output, and this would give users confidence in the system.

**Comment by JDWarner** [comment]
I've looked into this but it gets complicated fast, with even medium sized systems, especially for 3+ inputs and a bunch of AND connected Terms in Antecedents.

Opinions appreciated for how to solve this kind of problem efficiently.

Back on topic, Default values would be a reasonable stopgap for the short term, as I expect the ideal solution to require some thought. When a default value gets used, it would be nice to throw a Warning including the set of inputs, so people know it's happening and where the gap was.

**Comment by jcozar87** [comment]
How about set by default an output prediction if it has not been set, for instance the mode of the output/s? (so it would work with both regression and classification problems).

**Comment by JDWarner** [comment]
Yes, that's how it should work on creation. It would be nice to also allow a user to create a variable without specifying the default, then define one later.

**Comment by jcozar87** [comment]
Yes of course. You can create a ControlSystem without specifying the default value /then the mode will be used). Otherwise, it will use the set default prediction.

However, I am wondering if sometimes you need to know if the default prediction has been returned or not (for example, my belief in the prediction would depend on this). In this way, probably the option to raise an exception and manage it externally would be a better choice... what do you think?

**Comment by JDWarner** [comment]
Raising an exception is the current model. It's not a bad model, as you say, because the system truly doesn't know what to do with those inputs. Anything we assume imposes our assumptions on the model. I actually like raising exceptions, with the notable exception that right now we can't give anyone an explicit guarantee this _won't_ happen for a given system.

For the record, I do agree with @pbreach that this and a few other exceptions should be customized further than `ValueError` looking forward.

**Comment by jcozar87** [comment]
Yes, I totally agree with both of you. That's probably the best way to deal with it. Customizing the exception would allow us to distinguish between a non matching input and other type of errors.

**Comment by JDWarner** [comment]
:+1: for a customized exception. When I wrote this I was more focused on getting it working than making custom exceptions, but now would be an appropriate time.

I've been looking into possibly using a SAT solver (e.g., pycosat) as an option to analyze a given system given its Rules and connectivity, to determine if there are _any_ sets of inputs where this would happen. That kind of check would be very valuable! 

I think it's possible to do this with a SAT solver, but other commitments are taking my time and it won't go in before 0.2.
